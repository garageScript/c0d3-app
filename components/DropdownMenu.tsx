import React from 'react'
import DropdownButton from 'react-bootstrap/DropdownButton'
import Dropdown from 'react-bootstrap/Dropdown'
import styles from '../scss/dropDown.module.scss'
import { DropDirection } from 'react-bootstrap/esm/DropdownContext'

//a null item indicates a dropdown divider
export type Item = {
  title: string
  path?: string
  as?: 'a' | 'button'
  onClick?: Function
} | null

type DropDownMenuProps = {
  drop?: DropDirection
  items: Item[]
  title: string
  size?: 'sm' | 'lg' | undefined
  variant?:
    | 'primary'
    | 'secondary'
    | 'success'
    | 'info'
    | 'warning'
    | 'danger'
    | 'none'
  //changes the underlying component CSS base class name
  //https://react-bootstrap.github.io/components/dropdowns/#api
  bsPrefix?: string
}

export const DropdownMenu: React.FC<DropDownMenuProps> = ({
  drop = 'down',
  variant = 'none',
  title,
  size,
  items,
  bsPrefix
}) => {
  const menuItems = items.map((item: Item, itemsIndex: number) =>
    !item ? (
      <Dropdown.Divider key={itemsIndex} />
    ) : (
      <div className="text-center py-2 px-4" key={item.title}>
        <Dropdown.Item
          as={item.as || 'a'}
          key={itemsIndex}
          href={item.path}
          onClick={() => item.onClick && item.onClick(item.title)}
          bsPrefix={bsPrefix}
        >
          {item.title}
        </Dropdown.Item>
      </div>
    )
  )

  return (
    <>
      <div className="d-none d-lg-block">
        <DropdownButton
          title={title}
          variant={variant}
          size={size}
          drop={drop}
          bsPrefix={styles.title}
        >
          {menuItems}
        </DropdownButton>
      </div>
      <div className="d-lg-none">{menuItems}</div>
    </>
  )
}
