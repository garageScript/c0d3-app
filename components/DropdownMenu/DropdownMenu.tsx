import React, { useState } from 'react'
import Dropdown from 'react-bootstrap/Dropdown'
import styles from './dropdownMenu.module.scss'
import { DropDirection } from 'react-bootstrap/esm/DropdownContext'
import { ChevronRightIcon } from '@primer/octicons-react'

//a null item indicates a dropdown divider
export type Item = {
  title: string
  path?: string
  as?: 'a' | 'button'
  onClick?: Function
} | null

type DropDownMenuProps = {
  drop?: DropDirection
  items?: Item[] | null
  title?: string
  //changes the underlying component CSS base class name
  //https://react-bootstrap.github.io/components/dropdowns/#api
  bsPrefix?: string
  customToggle?: {
    Component?: () => JSX.Element
    style?: string
  }
}

const ChevronRight = () => <ChevronRightIcon size={17} />

export const DropdownMenu: React.FC<DropDownMenuProps> = ({
  items,
  title,
  customToggle,
  bsPrefix = ''
}) => {
  const [activeItem, setActiveItem] = useState({ title })

  return (
    <Dropdown bsPrefix={bsPrefix}>
      <Dropdown.Toggle
        bsPrefix={customToggle?.style ? customToggle.style : styles.dropdown}
        id="dropdown-lesson"
        data-testid="dropdown-lesson"
      >
        {customToggle?.Component ? (
          <customToggle.Component />
        ) : (
          <>
            {activeItem.title || 'None'}
            <ChevronRight />
          </>
        )}
      </Dropdown.Toggle>

      <Dropdown.Menu className={styles.dropdown__menu}>
        {items?.map((item, index) =>
          item ? (
            <Dropdown.Item
              key={`${item?.title}-${index}`}
              onClick={() => {
                item?.onClick?.(item)

                setActiveItem({
                  title: item?.title
                })
              }}
            >
              {item?.title}
            </Dropdown.Item>
          ) : (
            <Dropdown.Divider key={index} />
          )
        )}
      </Dropdown.Menu>
    </Dropdown>
  )
}
