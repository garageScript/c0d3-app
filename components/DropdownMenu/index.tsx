export { type Item, DropdownMenu } from './DropdownMenu'
