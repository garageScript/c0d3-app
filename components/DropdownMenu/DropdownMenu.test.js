import React from 'react'
import { render, fireEvent, screen, waitFor } from '@testing-library/react'
import { DropdownMenu } from './DropdownMenu'

// Imported to be able to use expect(...).toBeInTheDocument()
import '@testing-library/jest-dom'

const dropdownMenuItems = [
  { title: 'Lessons', path: '/admin/lessons', as: 'button' },
  null,
  { title: 'Users', path: '/admin/users', as: 'button' },
  { title: 'Alerts', path: '/admin/alerts', as: 'button' }
]

let testBtnOnClick = ''

describe('DropdownMenu Component', () => {
  test('Should render custom toggle', () => {
    const textToTest = 'custom toggle'

    const { queryByText } = render(
      <DropdownMenu
        items={dropdownMenuItems}
        customToggle={{ Component: () => <span>{textToTest}</span> }}
      />
    )

    expect(queryByText(textToTest)).toBeInTheDocument()
  })

  test('Should render custom toggle styles', () => {
    const { queryByText } = render(
      <DropdownMenu
        items={dropdownMenuItems}
        customToggle={{ style: 'customClass' }}
      />
    )

    expect(queryByText('None').classList.contains('customClass')).toBeTruthy()
  })

  test('Should render divider when an item is null', () => {
    const { container, queryByRole } = render(
      <DropdownMenu title="Admin" items={dropdownMenuItems} />
    )
    const div = queryByRole('separator')

    expect(div).toBeTruthy
    expect(container).toMatchSnapshot()
  })

  test('Should display default title when no title is passed', () => {
    const { queryByText } = render(<DropdownMenu items={dropdownMenuItems} />)

    expect(queryByText('None')).toBeInTheDocument()
  })

  test('Should change value of testBtnOnClick upon click', () => {
    dropdownMenuItems[0].onClick = val => (testBtnOnClick = val)

    const { container } = render(
      <DropdownMenu title="Admin" items={dropdownMenuItems} />
    )

    const btn = screen.queryByText('Admin')
    fireEvent.click(btn, { button: 1 })

    const lessons = screen.queryAllByText('Lessons')[0]
    fireEvent.click(lessons, { button: 1 })

    waitFor(() => expect(testBtnOnClick).toEqual('Lessons'))
    expect(container).toMatchSnapshot()
  })
})
