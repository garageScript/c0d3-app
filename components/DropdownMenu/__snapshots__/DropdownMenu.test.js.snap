// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu Component Should change value of testBtnOnClick upon click 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      class="dropdown btn btn-primary"
      data-testid="dropdown-lesson"
      id="dropdown-lesson"
      type="button"
    >
      Lessons
      <svg
        aria-hidden="true"
        class="octicon octicon-chevron-right"
        fill="currentColor"
        focusable="false"
        height="17"
        role="img"
        style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
        viewBox="0 0 16 16"
        width="17"
      >
        <path
          d="M6.22 3.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042L9.94 8 6.22 4.28a.75.75 0 0 1 0-1.06Z"
        />
      </svg>
    </button>
    <div
      aria-labelledby="dropdown-lesson"
      class="dropdown__menu dropdown-menu"
      x-placement="bottom-start"
    >
      <a
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="#"
        role="button"
        tabindex="0"
      >
        Lessons
      </a>
      <hr
        class="dropdown-divider"
        role="separator"
      />
      <a
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="#"
        role="button"
        tabindex="0"
      >
        Users
      </a>
      <a
        class="dropdown-item"
        data-rr-ui-dropdown-item=""
        href="#"
        role="button"
        tabindex="0"
      >
        Alerts
      </a>
    </div>
  </div>
</div>
`;

exports[`DropdownMenu Component Should render divider when an item is null 1`] = `
<div>
  <div
    class="dropdown"
  >
    <button
      aria-expanded="false"
      class="dropdown btn btn-primary"
      data-testid="dropdown-lesson"
      id="dropdown-lesson"
      type="button"
    >
      Admin
      <svg
        aria-hidden="true"
        class="octicon octicon-chevron-right"
        fill="currentColor"
        focusable="false"
        height="17"
        role="img"
        style="display: inline-block; user-select: none; vertical-align: text-bottom; overflow: visible;"
        viewBox="0 0 16 16"
        width="17"
      >
        <path
          d="M6.22 3.22a.75.75 0 0 1 1.06 0l4.25 4.25a.75.75 0 0 1 0 1.06l-4.25 4.25a.751.751 0 0 1-1.042-.018.751.751 0 0 1-.018-1.042L9.94 8 6.22 4.28a.75.75 0 0 1 0-1.06Z"
        />
      </svg>
    </button>
  </div>
</div>
`;
